src/folding/protogalaxy/folding.rs

```diff
@@ -0,0 +1,622 @@
+                }
+            }
+            let f_ev = eval_f(r1cs, &inner)?;
+
+            let mut Gsum = C::ScalarField::zero();
+            for (i, f_ev_i) in f_ev.iter().enumerate() {
+                let pow_i_betas = pow_i(i, &betas_star);
+                let curr = pow_i_betas * f_ev_i;
+                Gsum += curr;
+            }
+            G_evals[hi] = Gsum;
+        }
+        let G_X: DensePolynomial<C::ScalarField> =
+            Evaluations::<C::ScalarField>::from_vec_and_domain(G_evals.clone(), EH).interpolate();
+        let Z_X: DensePolynomial<C::ScalarField> = H.vanishing_polynomial().into();
+        // K(X) = (G(X)- F(alpha)*L_0(X)) / Z(X)
+        let L0_e = &L_X[0] * F_alpha; // L0(X)*F(a) will be 0 in the native case
+        let G_L0e = &G_X - &L0_e;
+        // Pending optimization: move division by Z_X to the prev loop
+        let (K_X, remainder) = G_L0e.divide_by_vanishing_poly(H).ok_or(
+            // using Error::ShouldNotReach because the method divide_by_vanishing_poly should never
+            // return 'None' despite returning an 'Option'.
+            Error::ShouldNotReach,
+        )?;
+        if !remainder.is_zero() {
+            return Err(Error::ProtoGalaxy(ProtoGalaxyError::RemainderNotZero));
+        }
+
+        transcript.absorb_vec(&K_X.coeffs);
+
+        let gamma = transcript.get_challenge();
+
+        let e_star =
+            F_alpha * L_X[0].evaluate(&gamma) + Z_X.evaluate(&gamma) * K_X.evaluate(&gamma);
+
+        let mut phi_star: C = instance.phi * L_X[0].evaluate(&gamma);
+        for i in 0..k {
+            phi_star += vec_instances[i].phi * L_X[i + 1].evaluate(&gamma);
+        }
+        let mut w_star: Vec<C::ScalarField> = vec_scalar_mul(&w.w, &L_X[0].evaluate(&gamma));
+        for i in 0..k {
+            w_star = vec_add(
+                &w_star,
+                &vec_scalar_mul(&vec_w[i].w, &L_X[i + 1].evaluate(&gamma)),
+            )?;
+        }
+        let mut r_w_star: C::ScalarField = w.r_w * L_X[0].evaluate(&gamma);
+        for i in 0..k {
+            r_w_star += vec_w[i].r_w * L_X[i + 1].evaluate(&gamma);
+        }
+
+        Ok((
+            CommittedInstance {
+                betas: betas_star,
+                phi: phi_star,
+                e: e_star,
+            },
+            Witness {
+                w: w_star,
+                r_w: r_w_star,
+            },
+            F_X_dense.coeffs,
+            K_X.coeffs,
+        ))
+    }
+
+    /// implements the non-interactive Verifier from the folding scheme described in section 4
+    pub fn verify(
+        transcript: &mut (impl Transcript<C> + ProtoGalaxyTranscript<C>),
+        r1cs: &R1CS<C::ScalarField>,
+        // running instance
+        instance: &CommittedInstance<C>,
+        // incomming instances
+        vec_instances: &[CommittedInstance<C>],
+        // polys from P
+        F_coeffs: Vec<C::ScalarField>,
+        K_coeffs: Vec<C::ScalarField>,
```

Comment:
I assume we pass the coeffs because we will be consuming them on each verification?
If not, and we plan to reuse, we should pass a reference.

Also wondering if we should pass instead of a Vec, a `SparsePolynomial` istead.

Reply:
Regarding the reusability, I think that we will not reuse them on each verification, since they will differ.
Regarding the SparsePolynomial: https://github.com/privacy-scaling-explorations/folding-schemes/pull/37#discussion_r1421505024